File: 2021_22 - Practicum in Object Oriented Programming/Olympic Games Statistics App/GUI App/Java/src/olympics/events/Event.java
package olympics.events;

import java.util.Objects;

public class Event {

	public enum TEventType {
		Team, Individual;
	};


	// ==========================================
	// Attributes:


	TEventType event_type = TEventType.Team;
	String name;
	int oly_games_id = -1;


    // ==========================================
	// Constructors:


	/**
	 * 
	 * @param event_type
	 * @param name
	 * @param oly_games_id - optional, -1 is default
	 */
	public Event(String event_type_str,
				 String name,
				 Integer oly_games_id)
	{
		this.name = name;

		if (oly_games_id != null)
			this.oly_games_id = oly_games_id;

		if (event_type_str.equals("Team")) {
			this.event_type = TEventType.Team;
		} else {
			this.event_type = TEventType.Individual;
		}
	}

	/**
	 * 
	 * @param event_type
	 * @param name
	 * @param oly_games_id - optional, -1 is default
	 */
	public Event(TEventType event_type,
				 String name,
				 Integer oly_games_id)
	{
		this.event_type = event_type;
		this.name = name;
		if (oly_games_id != null)
			this.oly_games_id = oly_games_id;
	}

	
	// ==========================================
	// Getters:

	
	public TEventType getType() {
		return event_type;
	}

	public String getName() {
		return name;
	}

	public int getOlympicGamesID() {
		return oly_games_id;
	}

	
	// ==========================================
	// Hashing:


	@Override
	public int hashCode() {
		return Objects.hash(event_type, name);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Event other = (Event) obj;
		return event_type == other.event_type
			&& Objects.equals(name, other.name);
	}
	
}

File: 2021_22 - Practicum in Object Oriented Programming/Olympic Games Statistics App/CLI App/src/Event.cpp
#include "Event.hpp"


// ================================
// Constructors:


sale::Event::Event(const sale::Event::TEventType event_type,
                   const std::string &name, const int oly_games_id)
    : event_type(event_type), name(new const std::string(name)),
      oly_games_id(oly_games_id)
{
}

sale::Event::Event(const TEventType event_type,
                   std::shared_ptr<const std::string> name_ptr,
                   const int oly_games_id)
    : event_type(event_type), name(std::move(name_ptr)),
      oly_games_id(oly_games_id)
{
}


// ================================
// HashableData overrides:


sale::Event::TKey sale::Event::getKey() const
{
    return this->name;
}


// ================================
// Getters:


sale::Event::TEventType sale::Event::getType() const
{
    return this->event_type;
}

std::string sale::Event::getName() const
{
    return *this->name;
}

const std::string& sale::Event::getNameReference() const
{
    return *this->name;
}

const int sale::Event::getOlympicGamesID() const
{
    return this->oly_games_id;
}


// ================================
// Operator Overloads:


bool operator==(const sale::Event& x, const sale::Event& y)
{
    if (x.event_type != y.event_type) return false;
    static auto key_cmp_func = sale::Event::KeyEqStruct();
    return key_cmp_func(x.name, y.name);
}
